src/ip/QosConfig.cc

```diff
@@ -132,6 +132,63 @@ void Ip::Qos::getNfmarkFromServer(const Comm::ConnectionPointer &server, const f
 #endif
 }
 
+nfmark_t Ip::Qos::getNfmarkFromClient(const Comm::ConnectionPointer &client)
```

Comment:
Most of this new code is exact duplicate of the getNfmarkFromServer() method. It would be better to make that existing method a more generic getNfMakeFromConnection(...) and have getNfmarkFromServer() and getNfmarkFromClient() specialize from that by sending the appropriate callback pointer to the generic shared method.

Once done the "nfmark" member of Comm::ConnectionPointer should be the remote endpoints CONNMARK, regardless of whether it is a server or client connection. So you can remove the new client_connmark member.

The debugs currently at line 121 (and new line 171), when in the generic method can be made to print the Comm::ConnectionPointer to self-document the connection instead of mentioning destination and source explicitly.

Please also take the opportunity to convert debugs about strerror(errno) to xstrerr(xerrno) construct on new and changed code lines. The debugs() macro uses time lookup syscalls which can change errno contents - so it must be preserved explicitly. See line 74-75 in this file for how to do it properly.